core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -304,20 +361,26 @@ private ThreadContextStruct copyHeaders(Iterable<Map.Entry<String, String>> head
         }
 
         private void writeTo(StreamOutput out, Map<String, String> defaultHeaders) throws IOException {
-            final Map<String, String> headers;
+            final Map<String, String> persistentHeaders;
             if (defaultHeaders.isEmpty()) {
-                headers = this.headers;
+                persistentHeaders = this.persistentHeaders;
             } else {
-                headers = new HashMap<>(defaultHeaders);
-                headers.putAll(this.headers);
+                persistentHeaders = new HashMap<>(defaultHeaders);
+                persistentHeaders.putAll(this.persistentHeaders);
             }
 
-            int keys = headers.size();
-            out.writeVInt(keys);
-            for (Map.Entry<String, String> entry : headers.entrySet()) {
+            out.writeVInt(persistentHeaders.size());
+            for (Map.Entry<String, String> entry : persistentHeaders.entrySet()) {
                 out.writeString(entry.getKey());
                 out.writeString(entry.getValue());
             }
+
+            out.writeMapOfLists(responseHeaders);
+        }
+
+        private void writeOnlyResponseHeadersTo(StreamOutput out) throws IOException {
```

Comment:
Can we fold this into `writeTo` and add a boolean to the signature (maybe `includeRequestHeaders`)? It seems like it just requires adding a if/else block